Dockerfile

```diff
@@ -1,49 +1,47 @@
 FROM ubuntu:24.04 AS builder
 LABEL org.opencontainers.image.authors="Kevin Kreiser <kevinkreiser@gmail.com>"
 
-ARG CONCURRENCY
-ARG ADDITIONAL_TARGETS
-ARG VERSION_MODIFIER=""
-
 # set paths
 ENV PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$PATH
 ENV LD_LIBRARY_PATH=/usr/local/lib:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/lib32:/usr/lib32
 RUN export DEBIAN_FRONTEND=noninteractive && apt update && apt install -y sudo
 
 # install deps
-WORKDIR /usr/local/src/valhalla
-COPY ./scripts/install-linux-deps.sh /usr/local/src/valhalla/scripts/install-linux-deps.sh
-RUN bash /usr/local/src/valhalla/scripts/install-linux-deps.sh
-RUN rm -rf /var/lib/apt/lists/*
+WORKDIR /src/valhalla
+COPY scripts/install-linux-deps.sh scripts/install-linux-deps.sh
+RUN bash ./scripts/install-linux-deps.sh
 
-# get the code into the right place and prepare to build it
-ADD . .
-RUN ls -la
-RUN git submodule sync && git submodule update --init --recursive
-RUN rm -rf build && mkdir build
+# get the code into the right place explicitly (for layer caching) and prepare to build it
+COPY cmake cmake/
+COPY date_time/ date_time/
+COPY locales/ locales/
+COPY lua/ lua/
+COPY proto/ proto/
+COPY scripts/ scripts/
+COPY src/ src/
+COPY third_party/ third_party/
+COPY valhalla/ valhalla/
+COPY CMakeLists.txt .
+COPY libvalhalla.pc.in .
+COPY LICENSE.md .
+COPY CHANGELOG.md .
+COPY COPYING .
+COPY README.md .
 
-# configure the build with symbols turned on so that crashes can be triaged
-WORKDIR /usr/local/src/valhalla/build
+ARG CONCURRENCY
+ARG ADDITIONAL_TARGETS
+# this arg will always compile the full project on every build bcs it's a commit hash
+# but we only build the docker image in GHA if relevant files changed (extensive paths-ignore in GHA workflow)
+ARG VERSION_MODIFIER=""
 # switch back to -DCMAKE_BUILD_TYPE=RelWithDebInfo and uncomment the block below if you want debug symbols
-RUN cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=gcc -DENABLE_SINGLE_FILES_WERROR=Off -DVALHALLA_VERSION_MODIFIER=${VERSION_MODIFIER}
-RUN make all ${ADDITIONAL_TARGETS} -j${CONCURRENCY:-$(nproc)}
-RUN make install
+RUN cmake -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=gcc -DENABLE_TESTS=Off -DENABLE_SINGLE_FILES_WERROR=Off -DVALHALLA_VERSION_MODIFIER=${VERSION_MODIFIER} && \
+  make -C build all ${ADDITIONAL_TARGETS} -j${CONCURRENCY:-$(nproc)} && \
+  make -C build install
 
-# we wont leave the source around but we'll drop the commit hash we'll also keep the locales
+# we wont leave the source around but we'll drop the commit hash and we'll also keep the locales
 WORKDIR /usr/local/src
-RUN cd valhalla && echo "https://github.com/valhalla/valhalla/tree/$(git rev-parse HEAD)" > ../valhalla_version
-RUN for f in valhalla/locales/*.json; do cat ${f} | python3 -c 'import sys; import json; print(json.load(sys.stdin)["posix_locale"])'; done > valhalla_locales
-RUN rm -rf valhalla
+RUN echo "$VERSION_MODIFIER" > /usr/local/valhalla_version && \
```

Comment:
small caveat: previously _any_ image build would get the valhalla commit hash, now it's only the case when `--build-arg VERSION_MODIFIER=xx` was specified